### PR TITLE
Support java.util.Optional*

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: drift-codec/src/test/java/io/airlift/drift/codec/AbstractThriftCodecManagerTest.java [hunk 1]
```diff
@@ -284,6 +288,92 @@ public void testArrays()
         testRoundTripSerialize(arrayField, new TCompactProtocol.Factory());
     }
 
+    @Test
+    public void testAllOptionalField()
+            throws Exception
+    {
+        OptionalField optionalField = new OptionalField();
+        optionalField.aBooleanOptional = Optional.of(true);
+        optionalField.aByteOptional = Optional.of((byte) -1);
```
Comment (electrum) [inline drift-codec/src/test/java/io/airlift/drift/codec/AbstractThriftCodecManagerTest.java]: Use different values for each integer type just to be safe

File: drift-codec/src/test/java/io/airlift/drift/codec/AbstractThriftCodecManagerTest.java [hunk 2]
```diff
@@ -284,6 +288,92 @@ public void testArrays()
...
+        optionalField.aStructOptional = Optional.of(new BonkField("message", 42));
+        optionalField.aEnumOptional = Optional.of(Fruit.BANANA);
+
+        optionalField.aOptionalDouble = OptionalDouble.of(87.6d);
+        optionalField.aOptionalInt = OptionalInt.of(Integer.MAX_VALUE - 10);
+        optionalField.aOptionalLong = OptionalLong.of(Long.MAX_VALUE - 20);
+
+        optionalField.aCustomEnumOptional = Optional.of(Letter.C);
+        optionalField.aListBooleanOptional = Optional.of(ImmutableList.of(true));
+        optionalField.aListByteOptional = Optional.of(ImmutableList.of((byte) -1));
```
Comment (electrum) [inline drift-codec/src/test/java/io/airlift/drift/codec/AbstractThriftCodecManagerTest.java]: Same here

File: drift-codec/src/test/java/io/airlift/drift/codec/OptionalStruct.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        aListLongOptional = Optional.empty();
+        aListDoubleOptional = Optional.empty();
+        aListStringOptional = Optional.empty();
+        aListStructOptional = Optional.empty();
+        aListEnumOptional = Optional.empty();
+        aListCustomEnumOptional = Optional.empty();
+    }
+
+    @ThriftConstructor
+    public OptionalStruct(Optional<Boolean> aBooleanOptional,
```
Comment (electrum) [inline drift-codec/src/test/java/io/airlift/drift/codec/OptionalStruct.java]: Wrap first arg

File: drift-codec/src/main/java/io/airlift/drift/codec/internal/builtin/OptionalIntThriftCodec.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public class OptionalIntThriftCodec
+        implements ThriftCodec<OptionalInt>
+{
+    @Override
+    public ThriftType getType()
+    {
+        return new ThriftType(ThriftType.DOUBLE, OptionalInt.class, OptionalInt.empty());
```
[the remark this answers is not in the document]
Comment (dain) [reply drift-codec/src/main/java/io/airlift/drift/codec/internal/builtin/OptionalIntThriftCodec.java]: The ThriftType is only used for IDL generation in the case of a custom codec

**Pull request merged**